src/graphQLUtils.js

```diff
@@ -0,0 +1,621 @@
+      return {
+        type: "event",
+        messageID: data.message_id,
+        threadID: threadID,
+        isGroup: threadType === "GROUP",
+        senderID: data.message_sender.id,
+        timestamp: data.timestamp_precise,
+        eventType: "video_call",
+        snippet: data.snippet,
+
+        // @Legacy
+        logMessageType: "other"
+      };
+    case "VoiceCallMessage":
+      return {
+        type: "event",
+        messageID: data.message_id,
+        threadID: threadID,
+        isGroup: threadType === "GROUP",
+        senderID: data.message_sender.id,
+        timestamp: data.timestamp_precise,
+        eventType: "voice_call",
+        snippet: data.snippet,
+
+        // @Legacy
+        logMessageType: "other"
+      };
+    case "GenericAdminTextMessage":
+      return {
+        type: "event",
+        messageID: data.message_id,
+        threadID: threadID,
+        isGroup: threadType === "GROUP",
+        senderID: data.message_sender.id,
+        timestamp: data.timestamp_precise,
+        snippet: data.snippet,
+        eventType: data.extensible_message_admin_text_type.toLowerCase(),
+        eventData: formatEventData(data.extensible_message_admin_text),
+
+        // @Legacy
+        logMessageType: utils.getAdminTextMessageType(
+          data.extensible_message_admin_text_type
+        ),
+        logMessageData: data.extensible_message_admin_text // Maybe different?
+      };
+    default:
+      return { error: "Don't know about message type " + data.__typename };
+  }
+}
+
+function graphQLBatch(ctx, defaultFuncs, query) {
+  // `queries` has to be a string. I couldn't tell from the dev console. This
+  // took me a really long time to figure out. I deserve a cookie for this.
+  var form = {
+    queries: JSON.stringify({
+      // @TODO Do we want to support multiple queries?
+      o0: query
+    })
+  };
+
+  return defaultFuncs
+    .post("https://www.facebook.com/api/graphqlbatch/", ctx.jar, form)
+    .then(utils.parseAndCheckLogin(ctx, defaultFuncs))
+    .then(function(resData) {
+      if (resData.error) {
+        throw resData;
+      }
+      // This returns us an array of things. The last one is the success /
+      // failure one.
+      // @TODO What do we do in this case?
+      if (resData[resData.length - 1].error_results !== 0) {
+        throw new Error("well darn there was an error_result");
+      }
+
+      return resData[0].o0.data;
+    });
+}
+
+function loadMessage(ctx, defaultFuncs, threadID, messageId) {
+  var query = {
+    // This doc_id was valid on March 26th 2018.
+    doc_id: "1801329719924418",
+    query_params: {
+      thread_and_message_id: {
+        thread_id: threadID,
+        message_id: messageId
+      }
+    }
+  };
+
+  return graphQLBatch(ctx, defaultFuncs, query)
+    .then(function(resData) {
+      // TODO@ Figure out how to get the correct thread type
+      return formatMessageGraphQLResponse(threadID, undefined, resData.message);
```

Comment:
Currently I use `undefined` for threadType, so `isGroup` flag will always be false. Is there an easy way to get threadType without making an extra network call?